### PR TITLE
[PLGN-391] - Microsoft Teams -adding in new actions for getting messages from a chat and creating a chat

### PR DESCRIPTION
**Pull request by rbowden-r7**
### Description

Describe the proposed changes:

  - to add in new actions to allow a user to create a new chat and an action to allow a user to fetch the last 50 messages from a  chat 
  - bumping validators from 0.17.0 to 0.21.0 (https://app.snyk.io/org/icon-fwb/project/3a82fe52-9600-4d20-ad51-84c401dfde82#issue-SNYK-PYTHON-VALIDATORS-6008990)

https://issues.corp.rapid7.com/browse/PLGN-391

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing

screenshot of testing to create chats and fetch messages has been added to the ticket - https://issues.corp.rapid7.com/browse/PLGN-391

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code
- [ ] 
new unit test have been added for the new actions 
```
Ran 31 tests in 0.016s

OK
```

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py [hunk 1]
```diff
@@ -227,6 +227,53 @@ def send_html_message(
     return message
 
 
+def create_chat(connection: insightconnect_plugin_runtime.connection, members: str, topic: str) -> bool:
+    """
+    Creates a chat for given users
+
+    :param connection: Object (insightconnect_plugin_runtime.connection)
+    :param members: String
+    :return: boolean
```
Comment (joneill-r7) [inline plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py:236]: wrong return type here and in `list_messages_from_chat`?

File: plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py [hunk 2]
```diff
@@ -351,6 +398,37 @@ def get_message_from_chat(  # noqa: C901
...
+    response = requests.get(message_url, headers=headers, params=params)  # nosec B113
+
+    try:
+        response.raise_for_status()
+    except Exception as error:
+        raise PluginException(preset=PluginException.Preset.BAD_REQUEST, data=error)
+    try:
+        messages = response.json().get("value", [])
+    except Exception as error:
+        raise PluginException(preset=PluginException.Preset.INVALID_JSON, data=error)
```
Comment (joneill-r7) [inline plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py:422]: would it be nicer to have:

``` 
try:
    response.raise_for_status()
    messages = response.json().get("value", [])
except HTTPError as http_error:
    raise PluginException(preset=PluginException.Preset.BAD_REQUEST, data=http_error)
except ValueError as value_error:
    raise PluginException(preset=PluginException.Preset.INVALID_JSON, data=value_error)
except Exception error:
    raise PluginException(preset=PluginException.Preset.UNKNOWN, data=error)
```

Comment (rbowden-r7) [reply plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py:422]: This has been updated along side a small changed to the unit test to correctly test the http request handling

File: plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py (hunk 1 above)
Comment (rbowden-r7) [reply plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py:236]: this has been updated to now correctly  use dict and list

File: plugins/microsoft_teams/unit_test/test_create_teams_chat.py
Comment (igorski-r7) [inline plugins/microsoft_teams/unit_test/test_create_teams_chat.py:1]: Please add schema validation to the tests

File: plugins/microsoft_teams/unit_test/test_list_messages_in_chat.py
Comment (igorski-r7) [inline plugins/microsoft_teams/unit_test/test_list_messages_in_chat.py:1]: Same here 😄

File: plugins/microsoft_teams/unit_test/test_create_teams_chat.py
Comment (rbowden-r7) [reply plugins/microsoft_teams/unit_test/test_create_teams_chat.py:1]: I have added schema validation to all unit tests

File: plugins/microsoft_teams/unit_test/test_list_messages_in_chat.py
Comment (rbowden-r7) [reply plugins/microsoft_teams/unit_test/test_list_messages_in_chat.py:1]: I have added schema validation to all unit tests

**Review by joneill-r7 — COMMENTED** [review]
Two minors, rest looks good!

File: plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py [hunk 3]
```diff
@@ -227,6 +227,53 @@ def send_html_message(
...
+
+    try:
+        result.raise_for_status()
+    except Exception as error:
+        raise PluginException(cause="Create chat failed.", assistance=result.text, data=error) from error
+
+    if not result.status_code == 201:
+        raise PluginException(cause="Create channel returned an unexpected result.", assistance=result.text)
+
+    return result.json()
```
[the remark this answers is not in the document]
Comment (rbowden-r7) [reply plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py]: A try/except has been added here

File: plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py [hunk 4]
```diff
@@ -227,6 +227,53 @@ def send_html_message(
     return message
 
 
+def create_chat(connection: insightconnect_plugin_runtime.connection, members: str, topic: str) -> dict:
+    """
+    Creates a chat for given users
+
+    :param connection: Object (insightconnect_plugin_runtime.connection)
+    :param members: String
```
[the remark this answers is not in the document]
Comment (rbowden-r7) [reply plugins/microsoft_teams/icon_microsoft_teams/util/teams_utils.py]: this has been updated to a string to reflect the correct type

**Review by igorski-r7 — APPROVED** [review]

**Review by joneill-r7 — APPROVED** [review]
lgtm

**Pull request merged**